### PR TITLE
feat: add local worker activation and manifest supervision

### PR DESCRIPTION
**Pull request by Sunrisea**
## Summary

This PR adds local remote-managed worker support to loongsuite-pilot and hardens the plugin-probe runtime package lifecycle.

Main changes:

- Add `loongsuite-pilot worker` commands for local worker instance management:
  - `connect`
  - `list`
  - `status`
  - `disconnect`
  - `delete`
- Add local worker activation in the orchestrator.
- Add `worker.manifest.json` supervision for plugin-probe packages:
  - start worker processes from runtime bundles
  - track PID/status/log paths
  - stop process groups on shutdown/disconnect
  - restart on failure according to manifest policy
- Add runtime option passthrough for worker manifests.
- Add safer plugin package replacement with staging + backup restore.
- Add OTLP resource attribute passthrough from hook records.
- Improve Claude Code transcript handling for resume/synthetic records.
- Add unit coverage for local workers, worker manifests, package replacement, resource attributes, and Claude transcript edge cases.

## Validation

Validated locally:

- `npm run typecheck`
- targeted unit tests:
  - `tests/unit/deployment/plugin-probe-strategy.test.ts`
  - `tests/unit/local-workers/local-worker-activation-service.test.ts`
  - `tests/unit/local-workers/instance-store.test.ts`
  - `tests/unit/local-workers/worker-cli.test.ts`
  - `tests/unit/hooks/shared/resource-context.test.mjs`
  - `tests/unit/flushers/otlp-trace-flusher/conversion.test.ts`
  - `tests/unit/hooks/claude-code/hook-processor.test.mjs`
  - `tests/unit/hooks/claude-code/transcript-parser.test.mjs`
  - `tests/unit/inputs/base/hook-record-transform.test.ts`

Validated on a remote Linux host using the open-source branch plus existing server-side old runtime bundle and `agents.d`:

- branch: `feat/remote-managed-local-workers-sync`
- commit: `4928735`
- `npm ci`
- `npm run build`
- `npm run typecheck`
- targeted unit tests: 45 passed
- OpenClaw local worker activation smoke test:
  - reused the existing server-side old OpenClaw runtime package and `agents.d`
  - worker reached `running`
  - `worker list` showed `worker-01`
  - runtime options were passed through:
    - `--plugin-install-scope local`
    - `--model-config-mode managed-runtime`
  - test worker was disconnected and stopped after validation

## Notes

This PR intentionally does not include internal AgentTeams runtime templates or bundled runtime tarballs. Runtime templates and bundles can be provided by deployment packaging or local installation separately.

### CODE REVIEW

File: src/flushers/otlp-trace-flusher.ts [hunk 1]
```diff
@@ -424,10 +445,83 @@ export class OtlpTraceFlusher extends BaseFlusher {
     const handler = new ExtendedTelemetryHandler({ tracerProvider: provider });
 
     state = { provider, handler, inMem };
-    this.agentConvertStates.set(agentType, state);
+    this.agentConvertStates.set(key, state);
```
Comment (ralf0131) [inline src/flushers/otlp-trace-flusher.ts:462]: **[Warning]** The `agentConvertStates` Map is keyed by `agentType|stableJson(resourceAttrs)`. Each unique combination creates a new `BasicTracerProvider` that is only cleaned up on `shutdown()`. If resource attributes vary across batches (e.g., different `agentteams.worker.name` values), the map grows unbounded during normal operation. Consider adding an LRU eviction or periodic cleanup of stale entries.

File: src/local-workers/local-worker-activation-service.ts [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  private readonly activeFingerprints = new Map<string, string>();
+  private timer: ReturnType<typeof setInterval> | null = null;
+  private watcher: FSWatcher | null = null;
+  private refreshing = false;
+
+  constructor(options: LocalWorkerActivationServiceOptions) {
+    this.dataDir = options.dataDir;
+    this.pilotDir = options.pilotDir;
+    this.definitions = options.definitions;
+    this.strategy = new PluginProbeStrategy(options.dataDir, options.pilotDir);
```
Comment (ralf0131) [inline src/local-workers/local-worker-activation-service.ts:43]: **[Info]** A separate `PluginProbeStrategy` instance is created here, independent of the one owned by `DeploymentManager`. Both hold their own `WorkerManifestSupervisor`. If both try to start/stop the same worker concurrently, there could be a race. Consider sharing the strategy instance or documenting the coordination contract.

File: src/local-workers/instance-store.ts [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+  return record ? Object.keys(record).find(key => key !== '') : undefined;
+}
+
+function readNumber(value: unknown): number | undefined {
+  return typeof value === 'number' && Number.isFinite(value) ? value : undefined;
+}
+
+function isAlive(pid: number): boolean {
+  try {
+    process.kill(pid, 0);
```
Comment (ralf0131) [inline src/local-workers/instance-store.ts:415]: **[Info]** PID-file-based liveness check via `process.kill(pid, 0)` has an inherent PID-reuse race: a recycled PID could produce a false-positive "running" result. This is a well-known limitation of PID file management and acceptable here, but worth noting for future hardening (e.g., writing a start-time stamp and verifying it).

File: src/deployment/plugin-probe-strategy.ts [hunk 4]
```diff
@@ -257,6 +343,63 @@ export class PluginProbeStrategy implements DeployStrategy {
...
+      }
+      logger.error('package acquire failed', { destDir: source.destDir, error: String(err) });
+      return false;
+    } finally {
+      await fs.rm(stagingDir, { recursive: true, force: true }).catch(() => {});
+    }
+  }
+
+  private async renamePath(source: string, target: string): Promise<void> {
+    await fs.rename(source, target);
```
Comment (ralf0131) [inline src/deployment/plugin-probe-strategy.ts]: **[Warning]** `renamePath` uses `fs.rename`, which fails across filesystem boundaries (EXDEV). If the staging directory and `destDir` are on different mount points, the atomic swap fails and the backup-restore path may not cover all failure modes. Consider adding a `cp -r` + `rm` fallback when `rename` throws `EXDEV`.

**Review by ralf0131 — COMMENTED** [review]
﻿### Summary
Adds local worker management CLI (`connect`/`list`/`status`/`disconnect`/`delete`), worker manifest supervision with restart policies, safer plugin-probe package replacement (staging + backup), OTLP resource-attribute passthrough, and Claude Code transcript resume/synthetic-record fixes. Well-structured, good test coverage (10 test files), backward-compatible type changes, and proper sensitive-field filtering. Overall: solid contribution with minor suggestions below.

### Findings
- **[Warning]** `src/flushers/otlp-trace-flusher.ts:448` — `agentConvertStates` Map keyed by `agentType|resourceAttrs` can grow unbounded if resource attributes vary; no eviction during normal operation.
- **[Warning]** `src/deployment/plugin-probe-strategy.ts:400` — `fs.rename` in `renamePath` fails on cross-filesystem (EXDEV) scenarios; staging and dest may be on different mounts.
- **[Info]** `src/local-workers/local-worker-activation-service.ts:44` — Separate `PluginProbeStrategy` instance from `DeploymentManager`'s; both manage workers independently.
- **[Info]** `src/local-workers/instance-store.ts:415` — PID-file liveness check has inherent PID-reuse race.
- **[Info]** `src/deployment/worker-manifest-supervisor.ts:417` — Process-group signaling relies on `detached: true`; platform assumption worth documenting.

### Suggestions
- For the `agentConvertStates` Map: consider an LRU cap or periodic sweep of entries whose last-use timestamp exceeds a threshold.
- For `renamePath`: wrap `fs.rename` in a try/catch and fall back to `fs.cp(source, target, {recursive: true})` + `fs.rm(source, {recursive: true, force: true})` on `EXDEV`.
- The transcript-parser changes for Claude Code resume/synthetic records are a good fix — the `fallbackTs` / `laterTimestamp` chain ensures all LLM calls get a valid `request_start_time`.

### Cross-repo Note
The `resourceAttributes` field added to `AgentActivityEntry` (`src/types/events.ts`) and the `resourceAttributeKeys` config option flow through to the OTLP flusher. If `alibaba/loongsuite-python` emits similar resource attributes, ensure the attribute key naming is consistent across both collectors.

---
*Automated review by github-manager-bot*

File: src/deployment/worker-manifest-supervisor.ts [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+      process.kill(pid, 0);
+      return true;
+    } catch {
+      return false;
+    }
+  }
+
+  private signalProcessGroup(pgid: number, signal: NodeJS.Signals): boolean {
+    try {
+      process.kill(-pgid, signal);
```
Comment (ralf0131) [inline src/deployment/worker-manifest-supervisor.ts:448]: **[Info]** `process.kill(-pgid, signal)` signals the entire process group. This relies on `detached: true` (line 170) making the child its own session leader, which works on Linux/macOS. Worth adding a brief comment documenting this platform assumption for future maintainers.

**Review by linrunqi08 — COMMENTED** [review]
## Code Review: 独立性分析与对已有功能的影响

### 整体评价

PR 引入了 local worker 管理功能，整体设计较合理。新增的核心模块（`instance-store.ts`、`worker-cli.ts`、`local-worker-activation-service.ts`、`worker-manifest-supervisor.ts`）之间职责划分清晰，测试覆盖也比较充分。

### 独立性分析

**新增模块（独立性好）**：
- `src/local-workers/` 目录下的 3 个文件是全新模块，互相依赖但与已有代码耦合度较低
- `src/deployment/worker-manifest-supervisor.ts` 作为新增的进程管理模块，仅被 `PluginProbeStrategy` 引用
- `assets/hooks/shared/resource-context.mjs` 是独立的工具模块

**对已有功能的影响点**：
1. **`plugin-probe-strategy.ts` — deploy 流程变更**：`deploy()` 从原来的"原地解压"改为"staging + backup + rename"，这影响所有 plugin-probe 类型的部署（不仅限于 worker）。好处是更安全（原子替换），但引入了新的失败模式
2. **`transcript-parser.mjs` — 时间戳隔离改为 per-promptId**：`lastToolResultTs` 从全局变量改为 per-promptId Map，修正了跨 turn 时间戳污染的问题。但这会改变已有多 turn + tool 场景下的 `request_start_time` 数值，可能影响下游 dashboard
3. **`hook-processor.mjs` — offset 更新时机**：把 `_next_transcript_offset` 的更新移到 `turns.length === 0` 检查之前，改变了空 turn 场景的行为
4. **`otlp-trace-flusher.ts` — convertState 键值变更**：从按 `agentType` 分组改为按 `agentType + resourceAttributes` 分组，对无 worker 的现有部署无影响（resourceAttributes 为空时 key 等价），但引入了潜在的内存增长问题

### 发现的问题

以下是具体的 inline comments，按严重程度排序。

File: src/deployment/worker-manifest-supervisor.ts [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+      child.once('error', err => {
+        void failStart(err);
+      });
+
+      if (!child.pid) {
+        await failStart(new Error('worker process did not expose a pid'));
+        return false;
+      }
+
+      child.unref();
```
Comment (linrunqi08) [inline src/deployment/worker-manifest-supervisor.ts:215]: **[Bug - Race Condition]** `child.on("exit", ...)` 注册在 `await fs.writeFile(paths.pid)` 和 `await this.writeStatus(...)` 之后。如果子进程在这两个 await 期间立即退出（比如 entrypoint 脚本不存在或权限错误），`exit` 事件会在没有监听器的情况下触发并丢失。

**后果**：PID 文件残留（指向已死进程），status 停留在 `running`，`runtimes` Map 条目永远不会被清理，`restartPolicy` 永远不会触发。

**建议**：在 `spawn()` 之后、任何 `await` 之前立即注册 `exit` 监听器，将退出信息暂存后再异步处理。

File: src/deployment/plugin-probe-strategy.ts [hunk 7]
```diff
@@ -42,6 +50,12 @@ export class PluginProbeStrategy implements DeployStrategy {
       return true;
     }
 
+    const shouldStartWorker = !def.localWorkerRuntime;
+    const hasWorkerManifest = await this.workerSupervisor.hasManifest(config.source.destDir);
+    if (shouldStartWorker && hasWorkerManifest && !await this.workerSupervisor.isWorkerRunning(config.source.destDir)) {
```
Comment (linrunqi08) [inline src/deployment/plugin-probe-strategy.ts:55]: **[Bug - 错误的 PID 路径解析]** `isWorkerRunning()` 在 `needsDeploy()` 中被调用时没有传入 `options`（默认为 `{}`）。如果 `worker.manifest.json` 的 `paths.pid` 使用了 `${instance:...}` 占位符（如 `${instance:stateDir}/worker.pid`），`expandInstanceValue` 会返回空字符串，导致 PID 路径解析为错误位置。

**后果**：`isWorkerRunning` 永远返回 `false`，即使 worker 正在运行，也会在每个检查周期触发不必要的完整重新部署（stop -> uninstall -> install -> start）。

**建议**：对于 `needsDeploy()` 场景，要么传递正确的 options，要么让 `isWorkerRunning` 在检测到未展开的占位符时回退到扫描 `.agent-worker/` 目录下的 PID 文件。

File: src/local-workers/local-worker-activation-service.ts [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        source: {
+          ...source,
+          destDir: bundleDir(this.dataDir, instance.id),
+        },
+      },
+    };
+  }
+
+  private buildManifestInstance(instance: LocalWorkerInstance): Record<string, string> {
+    return {
```
Comment (linrunqi08) [inline src/local-workers/local-worker-activation-service.ts:173]: **[Bug - PID 路径不一致]** `isInstanceWorkerAlive` 从 `stateDir(dataDir, instance.id)/worker.pid` 硬编码路径读取 PID，但 `WorkerManifestSupervisor` 写入的 PID 路径由 manifest 的 `paths.pid` 模板决定。

如果 manifest 定义了自定义 PID 路径（如 `.agent-worker/runtime/worker.pid` 而非 `${instance:stateDir}/worker.pid`），两者不一致。

**后果**：`isInstanceWorkerAlive` 读取不存在的文件，返回 `false`，`reconcile()` 每 5 秒触发一次完整的 stop + redeploy 循环，即使 worker 运行正常。

**建议**：复用 `WorkerManifestSupervisor.isWorkerRunning()` 并传入正确的 instance options，而不是硬编码 PID 路径。

File: src/flushers/otlp-trace-flusher.ts (hunk 1 above)
Comment (linrunqi08) [inline src/flushers/otlp-trace-flusher.ts:462]: **[Efficiency - 内存泄漏风险]** `agentConvertStates` 现在按 `agentType + stableJson(resourceAttributes)` 作为 key。每个唯一组合都会创建新的 `BasicTracerProvider` + `InMemorySpanExporter`，且永远不会被清理（仅在 shutdown 时释放）。

**后果**：如果 `resourceAttributes` 的值随时间变化（如 worker name 轮换、instance id 变更），Map 会无限增长，导致长期运行的 daemon 进程内存持续上升。

**建议**：
1. 添加 LRU 淘汰策略或定时清理不活跃的 convert state
2. 或者限制 projected resource attributes 的基数（只允许有限的已知 key）

File: src/deployment/plugin-probe-strategy.ts [hunk 9]
```diff
@@ -53,38 +67,70 @@ export class PluginProbeStrategy implements DeployStrategy {
...
+          def.id,
+          destDir,
+          this.buildScriptEnv(def.id),
+          {
+            instance: options.instance,
+            runtimeOptions: options.runtimeOptions,
+          },
+        );
+        if (!workerStarted) {
+          return { success: false, agentId: def.id, deployMode: 'plugin-probe', error: 'worker start failed' };
```
Comment (linrunqi08) [inline src/deployment/plugin-probe-strategy.ts]: **[Design - deploy 半成功状态]** `acquirePackageIntoDest` 成功替换了旧包，install 脚本也执行成功，但如果 `workerSupervisor.startIfPresent` 返回 `false`（如 command 不存在、端口冲突），`deploy()` 返回 `{success: false}`。

**后果**：
1. 新包已经安装到磁盘，但 deploy record 不会更新（因为 success=false）
2. 下次 `needsDeploy()` 检查时发现 hash 不匹配（record 未更新），触发又一次完整的 stop -> uninstall -> install -> start 循环
3. 如果 worker start 的失败是持久性的（如配置错误），会无限循环重部署

**建议**：考虑将包安装和 worker 启动的成功状态分开追踪，或在 worker start 失败时仍更新 deploy record（包已正确安装），将 worker 的重启交给 restartPolicy 或 activation service 处理。

File: assets/hooks/claude-code-hook-processor.mjs [hunk 10]
```diff
@@ -276,9 +280,9 @@ async function exportSession(state, stopReason) {
     await waitForTranscriptStable(transcriptPath, baseOffset);
   }
 
-  if (!parseResult || parseResult.turns.length === 0) return;
-
+  if (!parseResult) return;
```
Comment (linrunqi08) [inline assets/hooks/claude-code-hook-processor.mjs:360]: **[Compatibility - 行为变更影响已有会话]** 原代码：`if (!parseResult || parseResult.turns.length === 0) return;`，offset 不会在空 turn 时更新。新代码：先更新 offset 再检查 `turns.length === 0`。

这个变更对于 synthetic-only transcript（resume 场景）是正确的修复。但如果 stop hook 在 session 中途触发（用户输入已写入 transcript 但 assistant 尚未响应），旧代码会在下次调用时重新解析这些字节，新代码会跳过它们。

**风险**：对于被中断的 session，可能丢失部分 turn 数据。不过考虑到 stop hook 通常在 session 结束时触发，实际影响较小。建议在 PR description 中注明这个行为变更。

File: src/deployment/plugin-probe-strategy.ts (hunk 7 above)
Comment (Sunrisea) [reply src/deployment/plugin-probe-strategy.ts:55]: fixed

File: src/local-workers/local-worker-activation-service.ts [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+      });
+      this.watcher.on('error', err => {
+        logger.warn('local worker watch failed', { error: String(err) });
+      });
+    } catch (err) {
+      logger.warn('local worker watch unavailable', { error: String(err) });
+    }
+
+    const intervalMs = Number(process.env.LOONGSUITE_LOCAL_WORKER_SCAN_INTERVAL_MS) || DEFAULT_SCAN_INTERVAL_MS;
+    this.timer = setInterval(() => void this.refresh('poll'), intervalMs);
```
Comment (ralf0131) [inline src/local-workers/local-worker-activation-service.ts:63]: **[Warning]** This `setInterval` keeps the Node.js event loop alive if `stop()` is not reached (e.g. an unhandled early-exit path). `worker-manifest-supervisor.ts` already calls `.unref()` on its restart timer (line 353) — consider mirroring that here: `this.timer = setInterval(...); this.timer.unref();` for consistency and safer lifecycle.

File: src/deployment/worker-manifest-supervisor.ts [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+      // start() uses detached=true, so the child becomes the process-group leader on Linux/macOS.
+      process.kill(-pgid, signal);
+      return true;
+    } catch (err) {
+      if ((err as NodeJS.ErrnoException).code === 'ESRCH') return false;
+      throw err;
+    }
+  }
+
+  private async waitForExit(pid: number, timeoutMs: number): Promise<void> {
```
Comment (ralf0131) [inline src/deployment/worker-manifest-supervisor.ts:456]: **[Info]** `waitForExit` busy-polls with `process.kill(pid, 0)` every 100ms. Functionally correct for the graceful-shutdown path, but consider racing the child's `exit` event against a timeout `Promise` instead — avoids repeated signal-zero probes and is slightly more efficient. Non-blocking.

**Review by ralf0131 — COMMENTED** [review]
﻿### Summary
Re-review after new commits. Adds local worker instance management (`connect`/`list`/`status`/`disconnect`/`delete`) and a `worker.manifest.json` supervisor with restart policies — solid, well-tested feature. Bootstrap tokens are stored with `0o600` permissions, worker processes are spawned (not exec'd) and managed as detached process groups, and shutdown is wired into the orchestrator lifecycle. Overall the implementation is clean; the findings below are minor hardening suggestions, none blocking.

### Suggestions
- Add `.unref()` to the activation-service poll interval so an abnormal exit path cannot leave the process dangling.
- Consider an `exit`-event race instead of `kill(pid,0)` polling in `waitForExit`.
- Switch test `execSync` tar calls to `execFileSync` to avoid shell interpolation.

### Cross-repo Note
No API/protocol surface shared with `loongsuite-python`; the local-worker model is pilot-internal.

---
*Automated review by github-manager-bot*

### Additional notes (not anchored to a changed line)
- **[INFO]** `tests/unit/deployment/plugin-probe-strategy.test.ts:423` — **[Info]** `execSync` with string-interpolated paths is a shell-injection smell, even in tests. Prefer `execFileSync('tar', ['-czf', tarball, '-C', destDir, '.'], { stdio: 'ignore' })` to bypass the shell entirely. Low priority, test-only. _(line outside diff)_

File: assets/hooks/claude-code/transcript-parser.mjs [hunk 13]
```diff
@@ -187,18 +243,24 @@ export function parseClaudeTranscript(transcriptPath, byteOffset = 0) {
   const llmCalls = [];
   const conversationHistory = [];
   let prevCount = 0;
-  let lastToolResultTs = null;
+  const lastToolResultTsByPromptId = new Map();
+  const updateLastToolResultTs = (promptId, ts) => {
+    if (!ts) return;
+    const key = promptMapKey(promptId);
+    const prev = lastToolResultTsByPromptId.get(key);
+    if (!prev || ts > prev) lastToolResultTsByPromptId.set(key, ts);
```
Comment (fangxiu-wf) [inline assets/hooks/claude-code/transcript-parser.mjs:253]: **[Praise — 独立认领]** 这处 fix(`lastToolResultTs` 按 promptId 分桶 + 下游 `splitIntoTurns` 第 396-401 行的 per-llmCall fallback chain)**正好解决了我们最近发现的客户问题** — 多 turn 长会话中,中间某条 user record 缺 timestamp,导致后续 llm.request 的 `time_unix_nano="0"`,被下游 `util-genai` 误转成 Unix epoch (1970) 起点的怪异 LLM span(duration ≈ 56 年)。我们前几天还在内部讨论修法。

你们这版方案比我们当时设计的"扩展兜底链"更根本:
1. 消除了跨 promptId 污染(用 Map 隔离)
2. `normalizeRequestStart` 多了 candidate > response 反向时序的防御

可选建议:PR description 加一句这个 motivation,后续维护者(包括我们自己)会感谢。

File: assets/hooks/shared/resource-context.mjs [hunk 14]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2026 Alibaba Group Holding Limited
+// SPDX-License-Identifier: Apache-2.0
+
+const MAX_RESOURCE_FIELD_VALUE_LENGTH = 512;
+const SENSITIVE_FIELD_NAME_RE = /(?:TOKEN|SECRET|PASSWORD|KEY|CREDENTIAL|HEADER|COOKIE)/i;
```
Comment (fangxiu-wf) [inline assets/hooks/shared/resource-context.mjs]: **[Nit — DRY + 精度,non-blocking]** `SENSITIVE_FIELD_NAME_RE` 和 `src/flushers/otlp-trace-flusher.ts` 里的 `SENSITIVE_RESOURCE_KEY_RE` 是字字相同的正则,两处独立维护;建议提取到一个共享常量(或从本文件 export 出来给 flusher 复用)。

另外正则本身是子串匹配,可能误杀合规字段:`LOOKUP_KEY` / `HEADER_NAME` / `MONKEY_PATCH` 这类 metadata 会因含 `KEY` / `HEADER` 等被拒。当前 `DEFAULT_RESOURCE_ENV_FIELD_MAP` 只配了 `WORKER_NAME` / `INSTANCE_ID`,实际触发面为零;但未来扩展或用户自定义 `fieldMap` 会踩。

建议改成精确模式,如:
```js
/(^|_)(TOKEN|SECRET|PASSWORD|CREDENTIAL|COOKIE)(_|$)|^(API_KEY|API_HEADER)$/i
```

当前用法无害,可作为后续小改进。

File: assets/hooks/claude-code-hook-processor.mjs [hunk 15]
```diff
@@ -332,15 +336,21 @@ function buildTurnRecords(turn, turnIndex, sessionId, prevHash, userId, turnStop
   const traceId = generateTraceId();
   const entrySpanId = generateSpanId();
   const agentSpanId = generateSpanId();
+  const collectedResourceAttributes = collectResourceAttributesFromEnv(process.env, { agentId: AGENT_ID });
```
Comment (fangxiu-wf) [inline assets/hooks/claude-code-hook-processor.mjs]: **[Minor optimization — optional]** `collectResourceAttributesFromEnv(process.env, ...)` 在 `buildTurnRecords` 每次调用都重扫一遍 env。env 在 hook 进程生命周期内不会变(hook process 是 per-Stop spawn 的一次性进程),建议提到模块顶层算一次:

```js
const RESOURCE_ATTRIBUTES = collectResourceAttributesFromEnv(process.env, { agentId: AGENT_ID });
```

一个 session 通常调几次 `buildTurnRecords`,实际收益微乎其微,但代码更干净。

File: src/flushers/otlp-trace-flusher.ts [hunk 16]
```diff
@@ -411,23 +438,119 @@ export class OtlpTraceFlusher extends BaseFlusher {
...
 
-    state = { provider, handler, inMem };
-    this.agentConvertStates.set(agentType, state);
+    state = { provider, handler, inMem, active: 0 };
+    this.agentConvertStates.set(key, state);
+    this.evictConvertStates();
     return state;
   }
 
+  private evictConvertStates(): void {
```
Comment (fangxiu-wf) [inline src/flushers/otlp-trace-flusher.ts:467]: **[Edge case 附议]** 附议 @ralf0131 和 @linrunqi08 关于 `agentConvertStates` 无界增长的关切,`MAX_CONVERT_STATES` + LRU 的设计方向正确。

一个**未在他们评论里覆盖**的边界:`evictConvertStates` 通过 `[...entries].find(([, state]) => state.active === 0)` 找可踢的;如果 64 个 state **全部 active>0**(理论可能,例如同 batch 内对 64 个不同 worker 并发 convert),find 返回 undefined → 不踢 → 新进的 state 让 `agentConvertStates.size` 临时超过 64。

不致命(active 会很快下降,下个 evict 会清理),但严格说违反 cap。这是有意为之(优先正确性而非严格 cap)还是希望 await 等闲下来?当前在常规负载下无问题,只是确认设计意图,可能值得在 method 注释里说明。

File: assets/hooks/claude-code/transcript-parser.mjs [hunk 17]
```diff
@@ -24,6 +24,56 @@ import fs from 'node:fs';
...
+  if (!a) return b || null;
+  if (!b) return a;
+  const aMs = parseTimestampMs(a);
+  const bMs = parseTimestampMs(b);
+  if (aMs === null) return b;
+  if (bMs === null) return a;
+  return bMs > aMs ? b : a;
+}
+
+function normalizeRequestStart(candidate, responseTs) {
```
Comment (fangxiu-wf) [inline assets/hooks/claude-code/transcript-parser.mjs:68]: **[Doc nit]** `normalizeRequestStart` 里 candidate > response 时返回 responseTs 的逻辑挺巧妙(防御 transcript 异常顺序导致 request_start 比 response 还晚的逻辑矛盾),但函数名只说 "normalize",不直接看代码不容易理解为什么这么做。建议加一行 inline 注释说明 motivation:

```js
// Defense against transcript anomalies: a request_start_time later than
// its own response timestamp is logically impossible. Cap the candidate
// so the span doesn't get a negative duration in OTLP.
```

**Review by fangxiu-wf — COMMENTED** [review]
已读 ralf0131 + linrunqi08 的 review,绝大部分 worker 相关问题不在我熟悉范围内。

我专注 review 了 **claude-code 相关 4 个文件**(`claude-code-hook-processor.mjs` / `transcript-parser.mjs` / `shared/resource-context.mjs` / 间接相关的 `otlp-trace-flusher.ts`),结合已有评论补充几条独立观察。

**不点 APPROVE 不是因为 claude-code 部分有问题** — 它通过我所有测试和质量评估;留给熟悉 worker 模块的 reviewer 拍板整体合并。

---

一个 PR 流程性建议:**当前 PR base 在 `e6ac2b1`,早于 #67 合并到 main**。我本地 simulate `merge main` 后 typecheck ✓,关键测试 43/43 通过(`buildTurnRecords` 同时承载 intercept 和 resourceAttributes 两组逻辑,语义不冲突),所以不会 merge 失败,但 reviewer 看到的 diff 跟实际合并后状态有差距,建议 author rebase / merge main 后再 force-push 让 review 更直观。

**Review by linrunqi08 — APPROVED** [review]

**Review by ralf0131 — COMMENTED** [review]
﻿### Summary
Re-review after the two follow-up commits (`fix: address local worker review nits`, `fix: address claude trace review nits`). The previous findings are addressed well:

- `setInterval` now calls `.unref()` — the timer no longer keeps the event loop alive (Warning fixed).
- Resource attributes are collected once at module level instead of per-turn — avoids redundant parsing on every turn (perf improvement).
- The `agentConvertStates` eviction path is documented with its correctness rationale (temporary overflow preferred over dropping an active provider).

CLA is signed. Solid follow-up — thank you for iterating on the feedback.

### Findings
- **[Info]** `otlp-trace-flusher.ts` / `resource-context.mjs` — The sensitive-field regex was tightened from a loose substring match (`TOKEN|SECRET|PASSWORD|KEY|CREDENTIAL|HEADER|COOKIE`) to delimited-token matching plus exact `API_KEY`/`API_HEADER`. This correctly reduces false positives (e.g. a field like `KEYBOARD`), but note that `KEY` and `HEADER` are no longer matched as substrings at all — so a field such as `http.header.authorization` or `private.key` would no longer be redacted. If any mapped env var could surface such names, consider adding `authorization` or a delimited `HEADER` token back; otherwise the tradeoff is acceptable.

### Notes
- CI status is currently `UNSTABLE` — only the `license/cla` check is visible. As a first-time contributor, GitHub Actions workflows may require maintainer approval to run. Once CI is green this looks good to go.

---
*Automated review by github-manager-bot*

**Review by fangxiu-wf — APPROVED** [review]
